### DISCUSSION TITLE
How to get raw body of any response

### DISCUSSION THREAD

**Discussion by aceton41k** [Kong/unirest-java#415] [Q&A] [answered]
Hello! I want to receive raw body of any response (200, 400, 500) for using in unirest interceptor for logging.
In case of parsing error I receive ObjectResponse but have no idea how to get raw body.

**Comment by ryber** [comment, marked as answer]
If it's a parsing error the original body is kept on the exception ... which, I realize is a little weird. but it's like this: https://github.com/Kong/unirest-java/blob/f19ab4126e98289ab2434f19835f739e7558b0f2/unirest/src/test/java/BehaviorTests/AsObjectTest.java#L152-L163

**Reply by aceton41k** [reply]
> n ... which, I realize is a little weird. but it's like this:

Thanks, Ryan. I use similar case.
But I have noticed, if response returns empty body, I have not possobility to convert body to something else, if response becomes not empty. 
I will give an example.
Api request deletes message and if it's ok, returns code 200 and empty body. But also I need verify negative case (e.g. not authorized) which returns status 200 and body: {"code": 2}
My code use asEmpty() for positive case. But I couldn't get parsing error, if response returns body with code. This is bug or feature?
Should I use asObject in this cases?

**Reply by ryber** [reply]
A parsing error happens when the object mapper encounters an error in parsing the body. So with asEmpty(), we never even attempt, you could use something like asString(), then, if it's not null parse it into something else.

**Reply by aceton41k** [reply]
Ok, thank you!

**Comment by aceton41k** [comment]
Hello! How can I get rawbody of 500 response if no parsing error occured?
![image](https://user-images.githubusercontent.com/27342078/137511218-192c78c6-b777-4e88-b1a0-cde635e718f9.png)

**Comment by ryber** [comment]
There is a mapError method that will have the error. You can map it into a error object or just a string:
https://github.com/Kong/unirest-java/blob/main/unirest/src/test/java/BehaviorTests/ErrorParsingTest.java#L42-L62

**Reply by aceton41k** [reply]
Thank you, Ryan. mapError to string is that what I need
